internal/connectors/plugins/public/currencycloud/client/auth.go

```diff
@@ -20,36 +19,36 @@ func (c *Client) authenticate(ctx context.Context, httpClient *http.Client) (str
 	form.Add("login_id", c.loginID)
 	form.Add("api_key", c.apiKey)
 
-	req, err := http.NewRequest(http.MethodPost,
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
 		c.buildEndpoint("v2/authenticate/api"), strings.NewReader(form.Encode()))
 	if err != nil {
-		return "", fmt.Errorf("failed to create request: %w", err)
+		return fmt.Errorf("failed to create request: %w", err)
 	}
 
 	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
 	req.Header.Add("Accept", "application/json")
 
-	resp, err := httpClient.Do(req)
-	if err != nil {
-		return "", fmt.Errorf("failed to do get request: %w", err)
-	}
-
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		return "", unmarshalError(resp.StatusCode, resp.Body).Error()
-	}
-
 	//nolint:tagliatelle // allow for client code
 	type response struct {
 		AuthToken string `json:"auth_token"`
 	}
 
 	var res response
-
-	if err = json.NewDecoder(resp.Body).Decode(&res); err != nil {
-		return "", fmt.Errorf("failed to decode response body: %w", err)
+	var errRes currencyCloudError
+	_, err = c.httpClient.Do(req, &res, &errRes)
+	if err != nil {
+		return fmt.Errorf("failed to get authenticate: %w, %w", err, errRes.Error())
 	}
 
-	return res.AuthToken, nil
+	return nil
+}
+
+func (c *client) ensureLogin(ctx context.Context) error {
+	if c.authToken == "" {
```

Comment:
Actually I don't see where you set `c.authToken` anymore 🤔 .

Also you will need to protect `c.authToken` with a `sync.Mutex` since it can be set by multiple activities.

Reply:
The authToken is now set inside the method to authenticate, and this is protected against multiple write thanks to the singleflight 2 lines below normally !

Reply:
Oh nice. singleflight is new to me 🤓 

So shouldn't `c.authenticate()` contain some code where it sets `c.authToken = res.AuthToken` ? I don't see that happening anymore.

Reply:
I pushed again, i think it was truncated during the rebase, nice catch !

Reply:
LGTM now